mono_repo/lib/src/ci_shared.dart

```diff
@@ -180,5 +183,79 @@ void logPackages(Iterable<PackageConfig> configs) {
   }
 }
 
+/// Calculates the global stages ordering, and throws a [UserException] if it
+/// detects any cycles.
+///
+/// Ordering is determined by the order that stages appear in `mono_pkg.yaml`
+/// files, as well as `mono_repo.yaml` files if configured as conditional
+/// stages.
+///
+/// The [conditionalStages] are CI specific, as they use CI specific expression
+/// syntax.
+List<String> calculateOrderedStages(
+    RootConfig rootConfig, Map<String, ConditionalStage> conditionalStages) {
```

Comment:
@munificent I blame you for this lol

We need a lint or something you can enforce here instead of relying on reviews cause ill never do this by default lol.

Reply:
Agreed we need a lint!